sdk/core/allocator/main.cc

```diff
@@ -129,6 +129,92 @@ namespace
 		return (Capability{timeout}.is_valid() && timeout->may_block());
 	}
 
+	/**
+	 * Helper to reacquire the lock after sleeping.  This assumes that
+	 * `timeout` *was* valid but may have been freed and so checks only that it
+	 * has not been invalidated across yielding.  If `timeout` remains valid
+	 * then this adds any time passed as `elapsed` to the timeout and then
+	 * tries to reacquire the lock, blocking for no longer than the remaining
+	 * time on this timeout.
+	 *
+	 * This runs with interrupts disabled to ensure that the timeout remains
+	 * valid in between checking that it is valid and reacquiring the lock.
+	 * Once the lock is held, the timeout cannot be freed concurrently.
```

Comment:
While true at the moment, we have, in the past, discussed supporting multiple allocators.  Fortunately I think it doesn't much change the story?

Reply:
It makes it a bit of a mess, because the timeout can become invalid with the lock held, so there are other places where we need to be careful.  I don't really like the changes that multiple allocators would make to the programming model, I'm more inclined to think about supporting multiple heaps with a single allocator.